File: typescript/src/payment/domain/policies/paymentExpiration.ts
// Policy che verifica la scadenza di 48h e traduce PaymentDeadlineReached in ExpirePayment.

import type { Policy } from '@/common/domain/model/policy.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { ExpirePayment, expirePayment } from '../../application/commands/expirePayment.js';
import { PaymentDeadlineReached } from '../events/paymentDeadlineReached.js';
import type { Payment } from '../payment/payment.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

const DEADLINE_WINDOW_SECONDS = 48 * 3600;

export class PaymentExpiration implements Policy<PaymentDeadlineReached, ExpirePayment> {
  isDeadlineReached(payment: Payment, now: Timestamp): boolean {
    requireArgument(payment, 'payment');
    requireArgument(now, 'now');

    const deadline = payment.requestedAt().plusSeconds(DEADLINE_WINDOW_SECONDS);
    return now.isAfterOrEqual(deadline);
  }

  evaluate(event: PaymentDeadlineReached): ExpirePayment {
    requireArgument(event, 'event');
    return expirePayment(event.aggregateId);
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/policies/PaymentExpiration.java
package it.giannibombelli.wsc2026.payment.application.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.ExpirePayment;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentDeadlineReached;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;

public final class PaymentExpiration implements Policy<PaymentDeadlineReached, ExpirePayment> {

    private static final long DEADLINE_WINDOW_SECONDS = 48L * 3600;

    public boolean isDeadlineReached(Payment payment, Timestamp now) {
        Require.requireArgument(payment, "payment");
        Require.requireArgument(now, "now");

        Timestamp deadline = payment.requestedAt().plusSeconds(DEADLINE_WINDOW_SECONDS);
        return now.isAfterOrEqual(deadline);
    }

    @Override
    public ExpirePayment evaluate(PaymentDeadlineReached event) {
        Require.requireArgument(event, "event");
        return new ExpirePayment(event.aggregateId());
    }
}
